File: myCpps_2020/CF_CONTESTS/Round_658Div2/C2/CF_1382C2_PrefixFlipHardVersion.cpp
/*
-------------------------------------------------------------------
* @Name:           1382C1 Prefix Flip (Easy Version)
* @Author:         Shawn
* @Create Time:    2020/7/21 22:35:35  (UTC+08:00)
* @Url:            https://codeforces.com/contest/1382/problem/C1
* @Description:    
-------------------------------------------------------------------
                                     /~\
                                    |oo )
                                    _\=/_
                    ___        #   /  _  \
                   / ()\        \\//|/.\|\\
                 _|_____|_       \/  \_/  ||
                | | === | |         |\ /| ||
                |_|  O  |_|         \_ _/  #
                 ||  O  ||          | | |
                 ||__*__||          | | |
                |~ \___/ ~|         []|[]
                /=\ /=\ /=\         | | |
________________[_]_[_]_[_]________/_]_[_\_________________________
*/

#ifdef ONLINE_JUDGE
#define NDEBUG //ban assert when submit to online judge
#endif

#ifndef ONLINE_JUDGE
#include <chrono>
#endif

#include <bits/stdc++.h>

using namespace std;

using ill = long long;

class solution
{
  public:
    const int maxN = 1e5;

    void solve()
    {
        int tct; cin >> tct;
        for (int tcc = 1; tcc <= tct; ++tcc)
        {
            int n; cin >> n;
            string tmpA, tmpB; cin >> tmpA >> tmpB;
 
            vector<int> a(n, 0), b(n, 0);
            for (int i = 0; i <= n - 1; ++i) a[i] = (tmpA[i] == '1');
            for (int i = 0; i <= n - 1; ++i) b[i] = (tmpB[i] == '1');

            vector<int> ans0;
            for (int i = 0; i <= n - 1; ++i)
            {
                if (i == n - 1)
                {
                    if (a[i] == 1) ans0.push_back(i + 1);
                }
                else
                {
                    if (a[i] != a[i + 1]) ans0.push_back(i + 1);
                }
            }

            vector<int> ans1;
            for (int i = 0; i <= n - 1; ++i)
            {
                if (i == n - 1)
                {
                    if (b[i] == 1) ans1.push_back(i + 1);
                }
                else
                {
                    if (b[i] != b[i + 1]) ans1.push_back(i + 1);
                }
            }

            cout << ans0.size() + ans1.size() << '\n';
            for (auto x:ans0) cout << x << ' ';

            int siz = ans1.size();
            for (int i = siz - 1; i >= 0; --i)
            {
                if (i < siz - 1) cout << ' ';
                cout << ans1[i];
            }
            cout << '\n';
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);
    std::cout.tie(NULL);
#ifndef ONLINE_JUDGE
    freopen("CF_1382C2_PrefixFlipHardVersion.in", "r", stdin);
    freopen("CF_1382C2_PrefixFlipHardVersion.out", "w", stdout);
    auto startTime = std::chrono::high_resolution_clock::now();
#endif

    solution sln1;
    sln1.solve();
    cout.flush();

#ifndef ONLINE_JUDGE
    auto endTime = std::chrono::high_resolution_clock::now();
    std::cerr << "Execution time: "
              << std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime).count()
              << " ms\n";
#endif

    return 0;
}

File: myCpps_2020/CF_CONTESTS/Round_658Div2/C2/CF_1382C2_PrefixFlipHardVersion.java
import java.io.*;
import java.util.*;

public class CF_1382C2_PrefixFlipHardVersion
{
    public static void main(String[] args) throws IOException
    {
        boolean isLocal= System.getProperty("ONLINE_JUDGE") == null;
        long startTime = 0;
        if (isLocal)
        {
            startTime = System.currentTimeMillis();
            //redirect stdin/stdout to local file
            System.setIn(new FileInputStream(new File("CF_1382C2_PrefixFlipHardVersion.in")));
            System.setOut(new PrintStream(new File("CF_1382C2_PrefixFlipHardVersion.out")));
        }

        Scanner in = new Scanner(System.in);
        PrintStream out = System.out;

        new CF_1382C2_PrefixFlipHardVersion_Solution().Solve(in, out);

        in.close();
        out.close();
        
        if (isLocal)
        {
            System.err.println("program exited after: " + (System.currentTimeMillis() - startTime) + "ms");
        }        
    }
}


class CF_1382C2_PrefixFlipHardVersion_Solution
{
    public void Solve(Scanner In, PrintStream Out)
    {
        int tct = In.nextInt();
        for (int tcc = 1; tcc <= tct; ++tcc)
        {
            int n = In.nextInt();
            String tmpA = In.next();
            String tmpB = In.next();

            var a = new int[n];
            var b = new int[n];
            
            for (int i = 0; i <= n - 1; ++i) a[i] = ((tmpA.charAt(i) == '1') ? 1 : 0);
            for (int i = 0; i <= n - 1; ++i) b[i] = ((tmpB.charAt(i) == '1') ? 1 : 0);

            var ans0 = new ArrayList<Integer>();
            for (int i = 0; i <= n - 1; ++i)
            {
                if (i == n - 1)
                {
                    if (a[i] == 1) ans0.add(i + 1);
                }
                else
                {
                    if (a[i] != a[i + 1]) ans0.add(i + 1);
                }
            }

            var ans1 = new ArrayList<Integer>();
            for (int i = 0; i <= n - 1; ++i)
            {
                if (i == n - 1)
                {
                    if (b[i] == 1) ans1.add(i + 1);
                }
                else
                {
                    if (b[i] != b[i + 1]) ans1.add(i + 1);
                }
            }

            Out.println(ans0.size() + ans1.size());
            for (var x:ans0)
            {
                Out.print(x); Out.print(' ');
            }

            int siz = ans1.size();
            for (int i = siz - 1; i >= 0; --i)
            {
                if (i < siz - 1) Out.print(' ');
                Out.print(ans1.get(i));
            }
            Out.print('\n');
        }
    }
}
